hw/ip/spi_host/doc/_index.md

```diff
@@ -660,16 +659,613 @@ To use this reset, assert {{< regref "CONTROL.SW_RST" >}}, and then wait for the
+The FSM uses the `wait_cntr` register to ensure that it remains in this state for `csntrail+1` timeslices, during which time the active `csb` is still held low.
+The `wait_cntr` register resets to {{< regref "CONFIGOPT.CSNTRAIL" >}} upon entering this state, and is decremented once per timeslice.
+This state transitions to `WaitIdle` when `wait_cntr` is zero.
+
+6. WaitIdle state: In this timing control state, the FSM uses the `wait_cntr` register to ensure that all `csb` lines are held high for at least `csnidle+1` timeslices.
+The `wait_cntr` register resets to {{< regref "CONFIGOPT.CSNIDLE" >}} upon entering this state, and is decremented once per timeslice.
+This state transitions to `Idle` when `wait_cntr` reaches zero.
+
+{{< wavejson >}}
+{signal: [
+  {name: 'clk', wave: 'p...............'},
+  {name: 'rst_n', wave: '01..............'},
+  {name: 'state', wave: 'x22.42424242.2.2', data: ['Idle', 'WaitLead', 'IntClkHigh', 'IntClkLow', 'IntClkHigh', 'IntClkLow', 'IntClkHigh', 'IntClkLow','IntClkHigh', 'WaitTrail', 'WaitIdle', 'Idle']},
+  {name: 'csb (active device)', wave: 'x10..........1..'},
+  {name: 'csb (all others)', wave: '1...............'},
+  {name: 'sck', wave: '0...10101010....'}
+],
+ config: {hscale: 2}
+}
+{{< /wavejson >}}
+
+### Milestone Signals, Serial Data Lines & Shift Register Control
+
+The FSM manages I/O on the `sd` bus by controlling the timing of the shift register control signals: `shift_en_o`, `sample_en_o`, `rd_en_o`, `wr_en_o`, and `cmd_end_o`.
+
+The shift register control signals are managed through the use of three intermediate signals: 
+- `byte_starting`: This signal indicates the start of a new byte on the `sd` bus in the *following* clock cycle.
+For Bidirectional or TX segments this signal would indicate that it is time to load a new byte into the shift register.
+This signal corresponds to the FSM's `wr_en_o` port, though that output is suppressed during RX or dummy segments.
+- `byte_ending`: This signal indicates the end of the current `sd` byte in the *current* clock cycle (i.e., the next clock cycle either marks the beginning new byte or the end of the current segment).
+As illustrated in the following waveform, the `byte_starting` and `byte_ending` signals are often asserted at the same time, though there is an extra `byte_starting` pulse at the beginning of each command and an extra `byte_ending` pulse at the end.
+For RX and bidirectional command segments, a `byte_ending` pulse generates a `rd_en_o` pulse to the shift register, which transfers the 8-bit contents of the shift register into the RX FIFO via the Byte Merge block.
+- `bit_shifting`: This signal drives the `shift_en_o` control line to the shift register, ejecting the most-significant bits, and updating the `sd` outputs.
+
+These *milestone signals* mark the progress of each command segment.
+
+The coordination of the milestone signals and the shift register controls are shown in the following waveform.
+{{< wavejson >}}
```

Comment:
Are the arrows right here? They seem to skip backwards and forwards in time, which looks a bit odd to me.

Reply:
I agree it looks strange. However, the link between these signals is all combinational logic, and the causality is technically correct. The lower signal lines are driven high because the signals above have their values.    This all happens within the same clock cycle however, which is something particular that wavedrom seems not very helpful at illustrating.  Do you have any suggestions?

Reply:
Hi @rswarbrick, 
   Still not sure what do about the direction of the arrows, but I added a comment at the bottom. "NOTE: The milestone signals have a combinational dependence on the FSM state.".

Reply:
Hmm. For example, here:

![image](https://user-images.githubusercontent.com/104845/135603371-2768f3bd-4b6a-4c4d-b087-50f3cd046c42.png)

it looks like `byte_starting` going high depends on the fact that `state` will go to `WL` on the next cycle. I assume that's not quite right!

If the path is supposed to be `state = Idle -> byte_starting = 1 -> sample_en_o = 1` then I think you just need to move the `Z` label two dots to the left. Or is this supposed to mean something else?

Reply:
"it looks like byte_starting going high depends on the fact that state will go to WL on the next cycle. I assume that's not quite right!"
It *is* quite right!  That is exactly what is going on here.   It's magic, right?   Well no, not really, but I agree that it does make for a charmingly mysterious diagram.  

However thanks to your insistence, I now see how to make it clearer.

The issue is that FSM needs to instruct the shift register to do all these things as it *enters* the corresponding state.  So for example `byte_starting_cpha0` needs to be high as the FSM enters the WaitLead state.  The simplest way to express this is to peek at `state_d` whenever the `state_q` flops are enabled
 .
I can add `state_d` on the diagram, and then it will be clear that causality remains in effect in this IP.

Reply:
Hi @rswarbrick, what do you think if it after this latest commit?

Reply:
Ahah! That makes sense now! Yep, looks good to me (and no wibbly-wobbly timey-wimey stuff :-))

Reply:
That's a relief.   I've removed the timey-wimey wibbly-wobbly stuff from the following diagram as well.